File: tools/model_manifest.py
#!/usr/bin/env python3
"""Generate/verify Pixal3D GGUF model-set manifests.

The manifest is deliberately frontend-agnostic so Desktop and Web can share the
same version/size/SHA256 contract. It never downloads or deletes files.

Schema v1 intentionally describes the full textured production set. Semantic
constraints that JSON Schema cannot express for an array (unique names and the
complete required-role set) are enforced by validate_manifest_shape().
"""
from __future__ import annotations

import argparse
import hashlib
import json
import sys
import tempfile
from pathlib import Path

# flow 重みは model_family（mv = multiview, sv = single-view）ごとに別ファイル。
# decoder / image_encoder / naf は両 family で共通（同じバイト列を共有する）。
MODEL_FAMILIES = ("mv", "sv")
DEFAULT_MODEL_FAMILY = "mv"


def model_files(family: str = DEFAULT_MODEL_FAMILY) -> dict:
    if family not in MODEL_FAMILIES:
        raise ValueError(f"unknown model_family: {family!r}")
    return {
        "dinov3.gguf": {"role": "image_encoder", "required": True},
        "pixal3d_naf.gguf": {"role": "naf", "required": True},
        f"pixal3d_ss_flow_{family}.gguf": {"role": "ss_flow", "required": True},
        "ss_dec.gguf": {"role": "ss_decoder", "required": True},
        f"pixal3d_shape_flow_512_{family}.gguf": {"role": "shape_flow_512", "required": True},
        "shape_dec.gguf": {"role": "shape_decoder", "required": True},
        f"pixal3d_shape_flow_1024_{family}.gguf": {"role": "shape_flow_1024", "required": True},
        f"pixal3d_tex_flow_1024_{family}.gguf": {"role": "texture_flow_1024", "required": True},
        "tex_dec.gguf": {"role": "texture_decoder", "required": True},
    }


# 既存の MV manifest との互換のため MODEL_FILES は mv 集合を指したままにする。
MODEL_FILES = model_files(DEFAULT_MODEL_FAMILY)

ROOT_KEYS = {"schema_version", "model_set", "version", "model_family", "source", "files"}
FILE_KEYS = {"name", "role", "required", "size_bytes", "sha256"}


def sha256_file(path: Path) -> str:
    h = hashlib.sha256()
    with path.open("rb") as f:
        for chunk in iter(lambda: f.read(8 * 1024 * 1024), b""):
            h.update(chunk)
    return h.hexdigest()


def infer_family(files: list) -> str | None:
    """flow 4 role のファイル名末尾から family を推定する。混在・不明は None。"""
    flow_roles = {"ss_flow", "shape_flow_512", "shape_flow_1024", "texture_flow_1024"}
    names = [ent.get("name") for ent in files
             if isinstance(ent, dict) and ent.get("role") in flow_roles and isinstance(ent.get("name"), str)]
    if not names:
        return None
    for fam in MODEL_FAMILIES:
        if all(n.endswith(f"_{fam}.gguf") for n in names):
            return fam
    return None


def validate_manifest_shape(m: dict) -> list[str]:
    errors: list[str] = []
    if not isinstance(m, dict):
        return ["manifest must be an object"]

    unknown_root = sorted(set(m) - ROOT_KEYS)
    if unknown_root:
        errors.append("unknown top-level keys: " + ", ".join(unknown_root))
    if m.get("schema_version") != 1:
        errors.append("schema_version must be 1")
    if not isinstance(m.get("model_set"), str) or not m["model_set"]:
        errors.append("model_set missing")
    if not isinstance(m.get("version"), str) or not m["version"]:
        errors.append("version missing")
    if "source" in m and not isinstance(m["source"], str):
        errors.append("source must be a string")
    files = m.get("files")
    if not isinstance(files, list) or not files:
        return errors + ["files must be a non-empty array"]

    # family の決定は web/app/single_view.js modelFamilyForManifest と同じ規則:
    # flow のファイル名 (*_mv.gguf / *_sv.gguf) から推定し、model_family が明示されて
    # いればそれと一致しなければならない。混在・不明・不一致は拒否する。
    explicit = m.get("model_family")
    if explicit is not None and explicit not in MODEL_FAMILIES:
        errors.append("model_family must be one of: " + ", ".join(MODEL_FAMILIES))
        explicit = None
    inferred = infer_family(files)
    if explicit is not None and inferred is not None and explicit != inferred:
        errors.append(f"model_family {explicit} does not match the flow file names ({inferred})")
    family = explicit or inferred or DEFAULT_MODEL_FAMILY

    expected = model_files(family)
    seen_names: set[str] = set()
    seen_roles: set[str] = set()
    expected_roles = {spec["role"] for spec in expected.values() if spec["required"]}
    expected_names = {name for name, spec in expected.items() if spec["required"]}
    role_of_name = {name: spec["role"] for name, spec in expected.items()}
    name_of_role = {spec["role"]: name for name, spec in expected.items()}
    for i, ent in enumerate(files):
        p = f"files[{i}]"
        if not isinstance(ent, dict):
            errors.append(f"{p} must be object")
            continue
        unknown_file = sorted(set(ent) - FILE_KEYS)
        if unknown_file:
            errors.append(f"{p} has unknown keys: " + ", ".join(unknown_file))
        name = ent.get("name")
        if not isinstance(name, str) or not name or name in {".", ".."} or "/" in name or "\\" in name:
            errors.append(f"{p}.name invalid")
        elif name in seen_names:
            errors.append(f"duplicate file name: {name}")
        else:
            seen_names.add(name)
        role = ent.get("role")
        if not isinstance(role, str) or not role:
            errors.append(f"{p}.role invalid")
        elif role in seen_roles:
            errors.append(f"duplicate role: {role}")
        else:
            seen_roles.add(role)
        # name と role の対応を固定する（role の入れ替えで別種の GGUF を別段へ渡させない）。
        if isinstance(name, str) and isinstance(role, str):
            if name in role_of_name and role_of_name[name] != role:
                errors.append(f"{p}: {name} must have role {role_of_name[name]}, not {role}")
            elif name not in role_of_name and role in name_of_role:
                errors.append(f"{p}: role {role} must be {name_of_role[role]}, not {name}")
            if name in role_of_name and ent.get("required") is not True:
                errors.append(f"{p}: {name} must be required")
        if not isinstance(ent.get("required"), bool):
            errors.append(f"{p}.required must be bool")
        if not isinstance(ent.get("size_bytes"), int) or ent["size_bytes"] <= 0:
            errors.append(f"{p}.size_bytes invalid")
        digest = ent.get("sha256")
        if not isinstance(digest, str) or len(digest) != 64 or any(c not in "0123456789abcdef" for c in digest):
            errors.append(f"{p}.sha256 invalid")

    missing_names = sorted(expected_names - seen_names)
    missing_roles = sorted(expected_roles - seen_roles)
    if missing_names:
        errors.append("manifest missing required model files: " + ", ".join(missing_names))
    if missing_roles:
        errors.append("manifest missing required roles: " + ", ".join(missing_roles))
    return errors


def generate(root: Path, version: str, model_set: str, source: str | None,
             family: str = DEFAULT_MODEL_FAMILY) -> dict:
    files = []
    missing = []
    for name, spec in model_files(family).items():
        p = root / name
        if not p.is_file():
            if spec["required"]:
                missing.append(name)
            continue
        files.append({
            "name": name,
            "role": spec["role"],
            "required": spec["required"],
            "size_bytes": p.stat().st_size,
            "sha256": sha256_file(p),
        })
    if missing:
        raise SystemExit("missing required model files: " + ", ".join(missing))
    out = {"schema_version": 1, "model_set": model_set, "version": version, "files": files}
    if family != DEFAULT_MODEL_FAMILY:
        # 既存の mv manifest を変えないよう、mv 以外のときだけ明示する（省略時はファイル名から推定）。
        out["model_family"] = family
    if source:
        out["source"] = source
    return out


def verify(root: Path, manifest: dict) -> list[str]:
    errors = validate_manifest_shape(manifest)
    if errors:
        return errors
    for ent in manifest["files"]:
        p = root / ent["name"]
        if not p.is_file():
            if ent["required"]:
                errors.append(f"missing: {ent['name']}")
            continue
        size = p.stat().st_size
        if size != ent["size_bytes"]:
            errors.append(f"size mismatch {ent['name']}: {size} != {ent['size_bytes']}")
            continue
        digest = sha256_file(p)
        if digest != ent["sha256"]:
            errors.append(f"sha256 mismatch {ent['name']}")
    return errors


def self_test() -> dict:
    with tempfile.TemporaryDirectory() as td:
        root = Path(td)
        for i, name in enumerate(MODEL_FILES):
            (root / name).write_bytes((name + str(i)).encode())
        m = generate(root, "test", "pixal3d-test", None)
        assert verify(root, m) == []

        bad = json.loads(json.dumps(m))
        bad["unexpected"] = True
        assert any("unknown top-level" in e for e in validate_manifest_shape(bad))
        bad = json.loads(json.dumps(m))
        bad["files"][0]["unexpected"] = True
        assert any("unknown keys" in e for e in validate_manifest_shape(bad))
        bad = json.loads(json.dumps(m))
        bad["files"].pop()
        assert any("missing required" in e for e in verify(root, bad))
        bad = json.loads(json.dumps(m))
        bad["files"][1]["name"] = bad["files"][0]["name"]
        assert any("duplicate file name" in e for e in validate_manifest_shape(bad))
        bad = json.loads(json.dumps(m))
        bad["files"][0]["name"] = ".."
        assert any("name invalid" in e for e in validate_manifest_shape(bad))

        (root / "dinov3.gguf").write_bytes(b"corrupt")
        errs = verify(root, m)
        assert any("size mismatch dinov3.gguf" in e or "sha256 mismatch dinov3.gguf" in e for e in errs)

    # sv family: _sv の flow 名が要求され、mv の名前は満たさない。family の混在は拒否する。
    with tempfile.TemporaryDirectory() as td:
        root = Path(td)
        for i, name in enumerate(model_files("sv")):
            (root / name).write_bytes((name + str(i)).encode())
        sv = generate(root, "test", "pixal3d-sv-test", None, family="sv")
        assert sv["model_family"] == "sv"
        assert verify(root, sv) == []
        assert all(not n.endswith("_mv.gguf") for n in (f["name"] for f in sv["files"]))
        bad = json.loads(json.dumps(sv))
        del bad["model_family"]  # 省略時はブラウザと同じくファイル名から sv と推定される
        assert validate_manifest_shape(bad) == []
        bad = json.loads(json.dumps(sv))
        bad["model_family"] = "xx"
        assert any("model_family must be one of" in e for e in validate_manifest_shape(bad))
        bad = json.loads(json.dumps(m))
        bad["model_family"] = "sv"  # mv のファイル群に sv と書いても通らない
        assert any("does not match the flow file names" in e for e in validate_manifest_shape(bad))
        bad = json.loads(json.dumps(sv))
        bad["model_family"] = "mv"
        assert any("does not match the flow file names" in e for e in validate_manifest_shape(bad))
        # 混在（ss_flow だけ mv）は family 不明 → 既定 mv の集合を満たさず拒否
        bad = json.loads(json.dumps(sv))
        bad["files"][2]["name"] = "pixal3d_ss_flow_mv.gguf"
        assert any("missing required model files" in e for e in validate_manifest_shape(bad))
        # role の入れ替え（dinov3 <-> ss_flow）は名前と role の対応違反として拒否
        bad = json.loads(json.dumps(sv))
        bad["files"][0]["role"], bad["files"][2]["role"] = bad["files"][2]["role"], bad["files"][0]["role"]
        assert any("must have role" in e for e in validate_manifest_shape(bad))
        bad = json.loads(json.dumps(sv))
        bad["files"][0]["required"] = False
        assert any("must be required" in e for e in validate_manifest_shape(bad))
    print("MODEL_MANIFEST_SELF_TEST_OK")
    return m


def schema_test(schema_path: Path) -> None:
    try:
        import jsonschema
    except ImportError as exc:
        raise SystemExit("schema-test requires the 'jsonschema' package") from exc
    schema = json.loads(schema_path.read_text())
    jsonschema.Draft202012Validator.check_schema(schema)
    manifest = self_test()
    jsonschema.validate(instance=manifest, schema=schema)

    bad = json.loads(json.dumps(manifest))
    bad["unexpected"] = True
    try:
        jsonschema.validate(instance=bad, schema=schema)
    except jsonschema.ValidationError:
        pass
    else:
        raise AssertionError("schema unexpectedly accepted an unknown top-level property")
    bad = json.loads(json.dumps(manifest))
    bad["model_family"] = "xx"
    try:
        jsonschema.validate(instance=bad, schema=schema)
    except jsonschema.ValidationError:
        pass
    else:
        raise AssertionError("schema unexpectedly accepted an unknown model_family")
    for bad_name in (".", ".."):
        bad = json.loads(json.dumps(manifest))
        bad["files"][0]["name"] = bad_name
        try:
            jsonschema.validate(instance=bad, schema=schema)
        except jsonschema.ValidationError:
            pass
        else:
            raise AssertionError(f"schema unexpectedly accepted unsafe name {bad_name!r}")
    print("MODEL_MANIFEST_SCHEMA_TEST_OK")


def check_committed(models_root: Path, doc: Path) -> int:
    """Validate every committed release manifest and the digests recorded in the docs.

    実重みを持たない CI 用のゲート。manifest の形状・schema・9 role の充足と、
    release checklist に書かれた SHA256 が実ファイルと一致することだけを見る。
    """
    manifests = sorted(models_root.glob("*/pixal3d-models.json"))
    if not manifests:
        print(f"no committed manifest under {models_root}/*/pixal3d-models.json", file=sys.stderr)
        return 2
    failed = False
    doc_text = doc.read_text() if doc.is_file() else ""
    for mp in manifests:
        m = json.loads(mp.read_text())
        errors = validate_manifest_shape(m)
        expected_dir = f"{m.get('model_set')}-{m.get('version')}"
        if mp.parent.name != expected_dir:
            errors.append(f"directory {mp.parent.name} does not match {expected_dir}")
        digest = sha256_file(mp)
        if doc_text and digest not in doc_text:
            errors.append(f"sha256 {digest} is not recorded in {doc}")
        if errors:
            failed = True
            for e in errors:
                print(f"{mp}: {e}", file=sys.stderr)
        else:
            print(f"MODEL_MANIFEST_COMMITTED_OK {mp} {m['model_set']} {m['version']} {digest}")
    return 2 if failed else 0


def main() -> int:
    ap = argparse.ArgumentParser()
    sub = ap.add_subparsers(dest="cmd", required=True)
    g = sub.add_parser("generate")
    g.add_argument("models_dir", type=Path)
    g.add_argument("--version", required=True)
    g.add_argument("--model-set", default="pixal3d-q8_0")
    g.add_argument("--model-family", choices=MODEL_FAMILIES, default=DEFAULT_MODEL_FAMILY)
    g.add_argument("--source")
    g.add_argument("--output", type=Path, required=True)
    v = sub.add_parser("verify")
    v.add_argument("models_dir", type=Path)
    v.add_argument("manifest", type=Path)
    c = sub.add_parser("check-committed")
    c.add_argument("models_root", type=Path, nargs="?", default=Path("models"))
    c.add_argument("--doc", type=Path, default=Path("docs/PIXAL3D_RELEASE_CHECKLIST.md"))
    sub.add_parser("self-test")
    s = sub.add_parser("schema-test")
    s.add_argument("schema", type=Path, nargs="?", default=Path("models/pixal3d-model-set.schema.json"))
    a = ap.parse_args()

    if a.cmd == "self-test":
        self_test()
        return 0
    if a.cmd == "check-committed":
        return check_committed(a.models_root, a.doc)
    if a.cmd == "schema-test":
        schema_test(a.schema)
        return 0
    if a.cmd == "generate":
        m = generate(a.models_dir, a.version, a.model_set, a.source, family=a.model_family)
        a.output.write_text(json.dumps(m, indent=2) + "\n")
        print(a.output)
        return 0

    m = json.loads(a.manifest.read_text())
    errors = verify(a.models_dir, m)
    if errors:
        for e in errors:
            print(e, file=sys.stderr)
        return 2
    print(f"MODEL_MANIFEST_OK {m['model_set']} {m['version']}")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: src/model_manifest.cpp
#include "model_manifest.h"

#include <algorithm>
#include <cctype>
#include <cstdlib>
#include <filesystem>
#include <fstream>
#include <map>
#include <set>
#include <sstream>

namespace trellis {
namespace {

// ---- pixal3d-models.json だけを読む最小パーサ ------------------------------------------
// このリポジトリには汎用 JSON ライブラリが無い（transforms_json.cpp も同じ理由で自前）。
// 対象は「トップレベルが object、files が flat object の array」という既知の 1 形状だけ。
struct Json {
    enum class Kind { Null, Bool, Number, String, Array, Object } kind = Kind::Null;
    bool b = false;
    double num = 0.0;
    std::string str;
    std::vector<Json> arr;
    std::vector<std::pair<std::string, Json>> obj;

    const Json* find(const std::string& key) const {
        for (const auto& kv : obj) if (kv.first == key) return &kv.second;
        return nullptr;
    }
};

struct Parser {
    const std::string& s;
    size_t i = 0;
    std::string err;

    explicit Parser(const std::string& text) : s(text) {}

    void ws() { while (i < s.size() && (s[i] == ' ' || s[i] == '\t' || s[i] == '\n' || s[i] == '\r')) ++i; }
    bool lit(const char* text) {
        const size_t n = std::char_traits<char>::length(text);
        if (s.compare(i, n, text) != 0) return false;
        i += n;
        return true;
    }

    bool parse_string(std::string& out) {
        if (i >= s.size() || s[i] != '"') { err = "expected a string"; return false; }
        ++i;
        out.clear();
        while (i < s.size()) {
            const char c = s[i++];
            if (c == '"') return true;
            if (c != '\\') { out += c; continue; }
            if (i >= s.size()) break;
            const char e = s[i++];
            switch (e) {
                case '"': out += '"'; break;
                case '\\': out += '\\'; break;
                case '/': out += '/'; break;
                case 'b': out += '\b'; break;
                case 'f': out += '\f'; break;
                case 'n': out += '\n'; break;
                case 'r': out += '\r'; break;
                case 't': out += '\t'; break;
                case 'u': {
                    // manifest に非 ASCII は出ないが、来ても壊れないよう読み飛ばす。
                    if (i + 4 > s.size()) { err = "truncated \\u escape"; return false; }
                    i += 4;
                    out += '?';
                    break;
                }
                default: err = "invalid escape"; return false;
            }
        }
        err = "unterminated string";
        return false;
    }

    bool parse(Json& out) {
        ws();
        if (i >= s.size()) { err = "unexpected end of input"; return false; }
        const char c = s[i];
        if (c == '{') {
            ++i;
            out.kind = Json::Kind::Object;
            ws();
            if (i < s.size() && s[i] == '}') { ++i; return true; }
            while (true) {
                ws();
                std::string key;
                if (!parse_string(key)) return false;
                ws();
                if (i >= s.size() || s[i] != ':') { err = "expected ':'"; return false; }
                ++i;
                Json value;
                if (!parse(value)) return false;
                out.obj.emplace_back(std::move(key), std::move(value));
                ws();
                if (i < s.size() && s[i] == ',') { ++i; continue; }
                if (i < s.size() && s[i] == '}') { ++i; return true; }
                err = "expected ',' or '}'";
                return false;
            }
        }
        if (c == '[') {
            ++i;
            out.kind = Json::Kind::Array;
            ws();
            if (i < s.size() && s[i] == ']') { ++i; return true; }
            while (true) {
                Json value;
                if (!parse(value)) return false;
                out.arr.push_back(std::move(value));
                ws();
                if (i < s.size() && s[i] == ',') { ++i; continue; }
                if (i < s.size() && s[i] == ']') { ++i; return true; }
                err = "expected ',' or ']'";
                return false;
            }
        }
        if (c == '"') { out.kind = Json::Kind::String; return parse_string(out.str); }
        if (lit("true"))  { out.kind = Json::Kind::Bool; out.b = true;  return true; }
        if (lit("false")) { out.kind = Json::Kind::Bool; out.b = false; return true; }
        if (lit("null"))  { out.kind = Json::Kind::Null; return true; }
        {
            char* end = nullptr;
            const double d = std::strtod(s.c_str() + i, &end);
            if (end == s.c_str() + i) { err = "invalid value"; return false; }
            i = (size_t)(end - s.c_str());
            out.kind = Json::Kind::Number;
            out.num = d;
            return true;
        }
    }
};

bool ends_with(const std::string& s, const std::string& suffix) {
    return s.size() >= suffix.size() && s.compare(s.size() - suffix.size(), suffix.size(), suffix) == 0;
}

const char* const kFlowRoles[] = { "ss_flow", "shape_flow_512", "shape_flow_1024", "texture_flow_1024" };

bool is_flow_role(const std::string& role) {
    for (const char* r : kFlowRoles) if (role == r) return true;
    return false;
}

} // namespace

const char* model_family_name(ModelFamily f) { return f == ModelFamily::SV ? "sv" : "mv"; }

std::vector<std::pair<std::string, std::string>> required_files_for_family(ModelFamily f) {
    const std::string fam = model_family_name(f);
    return {
        { "dinov3.gguf",                              "image_encoder"     },
        { "pixal3d_naf.gguf",                         "naf"               },
        { "pixal3d_ss_flow_" + fam + ".gguf",         "ss_flow"           },
        { "ss_dec.gguf",                              "ss_decoder"        },
        { "pixal3d_shape_flow_512_" + fam + ".gguf",  "shape_flow_512"    },
        { "shape_dec.gguf",                           "shape_decoder"     },
        { "pixal3d_shape_flow_1024_" + fam + ".gguf", "shape_flow_1024"   },
        { "pixal3d_tex_flow_1024_" + fam + ".gguf",   "texture_flow_1024" },
        { "tex_dec.gguf",                             "texture_decoder"   },
    };
}

bool load_model_manifest(const std::string& path, ModelManifest& out, std::string& error) {
    std::ifstream f(path, std::ios::binary);
    if (!f) { error = "cannot read " + path; return false; }
    const std::string text((std::istreambuf_iterator<char>(f)), std::istreambuf_iterator<char>());

    Parser parser(text);
    Json root;
    if (!parser.parse(root) || root.kind != Json::Kind::Object) {
        error = "invalid pixal3d-models.json: " + (parser.err.empty() ? std::string("not an object") : parser.err);
        return false;
    }

    out = ModelManifest{};
    if (const Json* v = root.find("schema_version")) {
        if (v->kind != Json::Kind::Number) { error = "schema_version must be a number"; return false; }
        out.schema_version = (int)v->num;
    }
    if (out.schema_version != 0 && out.schema_version != 1) {
        error = "unsupported model manifest schema_version: " + std::to_string(out.schema_version);
        return false;
    }
    if (const Json* v = root.find("model_set")) { if (v->kind == Json::Kind::String) out.model_set = v->str; }
    if (const Json* v = root.find("version"))   { if (v->kind == Json::Kind::String) out.version = v->str; }

    // 1. 明示 family
    bool explicit_family = false;
    ModelFamily explicit_value = ModelFamily::MV;
    if (const Json* v = root.find("model_family")) {
        if (v->kind != Json::Kind::String || (v->str != "mv" && v->str != "sv")) {
            error = "model_family must be one of: mv, sv";
            return false;
        }
        explicit_family = true;
        explicit_value = (v->str == "sv") ? ModelFamily::SV : ModelFamily::MV;
    }

    const Json* files = root.find("files");
    if (!files || files->kind != Json::Kind::Array || files->arr.empty()) {
        error = "files must be a non-empty array";
        return false;
    }
    for (size_t k = 0; k < files->arr.size(); ++k) {
        const Json& ent = files->arr[k];
        if (ent.kind != Json::Kind::Object) {
            error = "files[" + std::to_string(k) + "] must be an object";
            return false;
        }
        ModelManifestFile mf;
        const Json* n = ent.find("name");
        const Json* r = ent.find("role");
        if (!n || n->kind != Json::Kind::String || n->str.empty()) {
            error = "files[" + std::to_string(k) + "].name invalid";
            return false;
        }
        if (!r || r->kind != Json::Kind::String || r->str.empty()) {
            error = "files[" + std::to_string(k) + "].role invalid";
            return false;
        }
        mf.name = n->str;
        mf.role = r->str;
        if (const Json* v = ent.find("sha256")) { if (v->kind == Json::Kind::String) mf.sha256 = v->str; }
        if (const Json* v = ent.find("size_bytes")) {
            if (v->kind != Json::Kind::Number) { error = mf.name + ": size_bytes must be a number"; return false; }
            mf.size_bytes = (long long)v->num;
        }
        if (const Json* v = ent.find("required")) { mf.required = (v->kind == Json::Kind::Bool) && v->b; }
        out.files.push_back(std::move(mf));
    }

    // 2. flow 4 role のファイル名から family を推定
    bool inferred_any = false;
    ModelFamily inferred = ModelFamily::MV;
    {
        std::vector<std::string> flow_names;
        for (const auto& mf : out.files) if (is_flow_role(mf.role)) flow_names.push_back(mf.name);
        if (!flow_names.empty()) {
            for (ModelFamily fam : { ModelFamily::MV, ModelFamily::SV }) {
                const std::string suffix = std::string("_") + model_family_name(fam) + ".gguf";
                if (std::all_of(flow_names.begin(), flow_names.end(),
                                [&](const std::string& s) { return ends_with(s, suffix); })) {
                    inferred_any = true;
                    inferred = fam;
                    break;
                }
            }
        }
    }

    // 3. 明示と推定の不一致は拒否
    if (explicit_family && inferred_any && explicit_value != inferred) {
        error = std::string("model_family ") + model_family_name(explicit_value) +
                " does not match the flow file names (" + model_family_name(inferred) + ")";
        return false;
    }
    // 4. family の確定
    out.family_explicit = explicit_family;
    out.family = explicit_family ? explicit_value : (inferred_any ? inferred : ModelFamily::MV);

    // 5-6. name<->role の固定、重複の拒否
    const auto expected = required_files_for_family(out.family);
    std::map<std::string, std::string> role_of_name;
    std::map<std::string, std::string> name_of_role;
    for (const auto& kv : expected) {
        role_of_name[kv.first] = kv.second;
        name_of_role[kv.second] = kv.first;
    }
    std::set<std::string> seen_names, seen_roles;
    for (size_t k = 0; k < out.files.size(); ++k) {
        const ModelManifestFile& mf = out.files[k];
        const std::string at = "files[" + std::to_string(k) + "]";
        if (!seen_names.insert(mf.name).second) { error = "duplicate file name: " + mf.name; return false; }
        if (!seen_roles.insert(mf.role).second) { error = "duplicate role: " + mf.role; return false; }
        const auto rn = role_of_name.find(mf.name);
        if (rn != role_of_name.end()) {
            if (rn->second != mf.role) {
                error = at + ": " + mf.name + " must have role " + rn->second + ", not " + mf.role;
                return false;
            }
            if (!mf.required) { error = at + ": " + mf.name + " must be required"; return false; }
        } else {
            const auto nr = name_of_role.find(mf.role);
            if (nr != name_of_role.end()) {
                error = at + ": role " + mf.role + " must be " + nr->second + ", not " + mf.name;
                return false;
            }
        }
    }

    // 7. 必須 9 name / 9 role が揃う
    std::vector<std::string> missing_names, missing_roles;
    for (const auto& kv : expected) {
        if (!seen_names.count(kv.first)) missing_names.push_back(kv.first);
        if (!seen_roles.count(kv.second)) missing_roles.push_back(kv.second);
    }
    if (!missing_names.empty()) {
        std::ostringstream os;
        os << "manifest missing required model files:";
        for (const auto& n : missing_names) os << " " << n;
        error = os.str();
        return false;
    }
    if (!missing_roles.empty()) {
        std::ostringstream os;
        os << "manifest missing required roles:";
        for (const auto& r : missing_roles) os << " " << r;
        error = os.str();
        return false;
    }
    return true;
}

ModelSetStatus inspect_model_set(const std::string& dir, ModelFamily expected) {
    ModelSetStatus st;
    st.family = model_family_name(expected);
    if (dir.empty()) {
        st.reason = std::string("no ") + model_family_name(expected) + " model directory is configured";
        return st;
    }
    st.configured = true;

    std::error_code ec;
    if (!std::filesystem::is_directory(dir, ec) || ec) {
        st.reason = "not a directory: " + dir;
        return st;
    }
    const std::string manifest_path = (std::filesystem::path(dir) / "pixal3d-models.json").string();
    if (!std::filesystem::exists(manifest_path, ec) || ec) {
        st.reason = "pixal3d-models.json is missing (an incomplete or unverified model directory)";
        return st;
    }

    ModelManifest manifest;
    std::string err;
    if (!load_model_manifest(manifest_path, manifest, err)) {
        st.reason = err;
        return st;
    }
    st.model_set = manifest.model_set;
    st.version = manifest.version;
    st.family = model_family_name(manifest.family);

    if (manifest.family != expected) {
        st.reason = std::string("model_family mismatch: this endpoint needs ") +
                    model_family_name(expected) + ", the manifest says " + st.family;
        return st;
    }

    for (const auto& mf : manifest.files) {
        if (!mf.required) continue;
        const std::filesystem::path p = std::filesystem::path(dir) / mf.name;
        if (!std::filesystem::exists(p, ec) || ec) {
            st.reason = "missing model file: " + mf.name;
            return st;
        }
        if (mf.size_bytes >= 0) {
            const auto actual = std::filesystem::file_size(p, ec);
            if (ec) { st.reason = "cannot stat model file: " + mf.name; return st; }
            if ((long long)actual != mf.size_bytes) {
                st.reason = "size mismatch " + mf.name + ": " + std::to_string((long long)actual) +
                            " != " + std::to_string(mf.size_bytes);
                return st;
            }
        }
    }

    st.available = true;
    return st;
}

} // namespace trellis
